pipeline/git.go

```diff
@@ -148,6 +164,100 @@ func updateAllCurrentPipelines() {
 	wg.Wait()
 }
 
+// GithubRepoService is an interface defining the Wrapper Interface
+// needed to test the github client.
+type GithubRepoService interface {
+	CreateHook(ctx context.Context, owner, repo string, hook *github.Hook) (*github.Hook, *github.Response, error)
+}
+
+// GithubClient is a client that has the ability to replace the actual
+// git client.
+type GithubClient struct {
+	Repositories GithubRepoService
+	*github.Client
+}
+
+// NewGithubClient creates a wrapper around the github client. This is
+// needed in order to decouple gaia from github client to be
+// able to unit test createGithubWebhook and ultimately have
+// the ability to replace github with anything else.
+func NewGithubClient(httpClient *gohttp.Client, repoMock GithubRepoService) GithubClient {
+	if repoMock != nil {
+		return GithubClient{
+			Repositories: repoMock,
+		}
+	}
+	client := github.NewClient(httpClient)
+
+	return GithubClient{
+		Repositories: client.Repositories,
+	}
+}
+
+func createGithubWebhook(token string, repo *gaia.GitRepo, gitRepo GithubRepoService) error {
+	vault, err := services.VaultService(nil)
+	if err != nil {
+		gaia.Cfg.Logger.Error("unable to initialize vault: " + err.Error())
+		return err
+	}
+
+	err = vault.LoadSecrets()
+	if err != nil {
+		gaia.Cfg.Logger.Error("unable to open vault: " + err.Error())
+		return err
+	}
+
+	ctx := context.Background()
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+	tc := oauth2.NewClient(ctx, ts)
+	config := make(map[string]interface{})
+	config["url"] = gaia.Cfg.Hostname + "/api/" + gaia.APIVersion + "/pipeline/githook"
+	secret, err := vault.Get("GITHUB_WEBHOOK_SECRET")
+	if err != nil {
+		secret = []byte(generateWebhookSecret())
+		vault.Add("GITHUB_WEBHOOK_SECRET", secret)
+		err = vault.SaveSecrets()
+		if err != nil {
+			return err
+		}
+	}
+	config["secret"] = string(secret)
+	config["content_type"] = "json"
+
+	client := NewGithubClient(tc, gitRepo)
+	repoName := path.Base(repo.URL)
+	repoName = strings.TrimSuffix(repoName, ".git")
+	// var repoLocation string
+	re := regexp.MustCompile("^(https|git)(:\\/\\/|@)([^\\/:]+)[\\/:]([^\\/:]+)\\/(.+)$")
+	m := re.FindAllStringSubmatch(repo.URL, -1)
+	if m == nil {
+		return errors.New("failed to extract url parameters from git url")
+	}
+	repoUser := m[0][4]
+	hook, resp, err := client.Repositories.CreateHook(context.Background(), repoUser, repoName, &github.Hook{
+		Events: []string{"push"},
+		Name:   github.String("web"),
+		Active: github.Bool(true),
+		Config: config,
+	})
+	if err != nil {
+		gaia.Cfg.Logger.Error("error while trying to create webhook: ", "error", err.Error())
+		return err
+	}
+	gaia.Cfg.Logger.Info("hook created: ", github.Stringify(hook.Name), resp.Status)
+	gaia.Cfg.Logger.Info("hook url: ", hook.GetURL())
```

Comment:
should be `gaia.Cfg.Logger.Info("hook url ", "url", hook.GetURL())`
Otherwise we get ugly log output like `hook url: : asdf=<unknown>` 😄 
It's always (initial_log_message, key=string, value)